src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,705 @@
+--
+{-# INLINE enumerateFromNum #-}
+enumerateFromNum :: (Monad m, Num a) => Unfold m a a
+enumerateFromNum = lmap (\from -> (from, 1)) enumerateFromStepNum
+
+-- | Generate a finite stream starting from a from and next values with
+-- stride of (next-from) up to value. The implementation is numerically
+-- stable for floating point values.
+--
+-- Note that 'enumerateFromThenToIntegral' is faster for integrals.
+--
+-- /Internal/
+--
+{-# INLINE enumerateFromThenToNum #-}
+enumerateFromThenToNum :: (Monad m, Ord a, Fractional a) =>
+    Unfold m (a, a, a) a
+enumerateFromThenToNum = Unfold step inject
+
+    where
+
+    inject (!from, !next, !to) = return (from, next - from, to, 0)
+
+    {-# INLINE_LATE step #-}
+    step (from, stride, to, !i) = return $
+        if stride == 0
+        then if from <= to then Yield from $! (from, stride, to, i + 1) else Stop
+        else
+            if stride > 0
+            then
+                if (from + i * stride) <= to + stride / 2
+                then (Yield $! (from + i * stride)) $! (from, stride, to, i + 1)
+                else Stop
+            else
+                if (from + i * stride) >= to + stride / 2
+                then (Yield $! (from + i * stride)) $! (from, stride, to, i + 1)
+                else Stop
+
+-- | Generate a finite stream starting from a from  till to value.
+-- stride of value 1 will be used. The implementation is numerically
+-- stable for floating point values.
+--
+-- Note that 'enumerateFromToIntegral' is faster for integrals.
+--
+-- /Internal/
+--
+{-# INLINE enumerateFromToNum #-}
+enumerateFromToNum :: (Monad m, Ord a, Fractional a) =>
+    Unfold m (a, a) a
+enumerateFromToNum = Unfold step inject
+
+    where
+
+    inject (!from, !to) = return (from, 1, to, 0)
+
+    {-# INLINE_LATE step #-}
+    step (from, !stride, to, !i) = return $
+            if (from + i * stride) <= to + stride / 2
+            then (Yield $! (from + i * stride)) $! (from, stride, to, i + 1)
+            else Stop
+
+------------------------------------------------------------------------------
+-- Enumeration of Integrals
+------------------------------------------------------------------------------
+
+-- | Can be used to enumerate unbounded integrals. This does not check for
+-- overflow or underflow for bounded integrals.
+--
+-- /Internal/
+{-# INLINE_NORMAL enumerateFromStepIntegral #-}
+enumerateFromStepIntegral :: (Integral a, Monad m) => Unfold m (a, a) a
+enumerateFromStepIntegral = Unfold step inject
+    where
+    inject (from, next) = from `seq` next `seq` return (from, next - from)
+    {-# INLINE_LATE step #-}
+    step (x, stride) = return $ Yield x $! (x + stride, stride)
+
+-- | Can be used to enumerate unbounded integrals. Starting with value from
+-- with stride of next-from till value of to.
+--
+-- /Internal/
+--
+{-# INLINE_NORMAL enumerateFromToStepIntegral #-}
+enumerateFromToStepIntegral :: (Integral a, Monad m) =>
+    Unfold m (a, a, a) a
```

Comment:
The name is incorrect. This function is actually `enumerateFromThenToIntegral`  because it takes `from then to` as arguments.